datasafe-storage/datasafe-storage-impl-s3/src/main/java/de/adorsys/datasafe/storage/impl/s3/MultipartUploadS3StorageOutputStream.java

```diff
@@ -0,0 +1,219 @@
+
+        MessageDigest messageDigest = MessageDigest.getInstance("MD5");
+        String md5Digest = BinaryUtils.toBase64(messageDigest.digest(content));
+        objectMetadata.setContentMD5(md5Digest);
+
+        amazonS3.putObject(
+                bucketName,
+                objectName,
+                new ByteArrayInputStream(content),
+                objectMetadata);
+
+        // Release the memory
+        currentOutputStream = null;
+    }
+
+    private void finishMultiPartUpload() throws IOException {
+        completionService.submit(
+                new UploadChunkResultCallable(ChunkUploadRequest.builder()
+                        .amazonS3(amazonS3)
+                        .content(currentOutputStream.toByteArray())
+                        .contentSize(currentOutputStream.size())
+                        .bucketName(bucketName)
+                        .objectName(objectName)
+                        .uploadId(multiPartUploadResult.getUploadId())
+                        .chunkNumberCounter(partCounter)
+                        .lastChunk(true)
+                        .build()
+                )
+        );
+
+        try {
+            List<PartETag> partETags = getMultiPartsUploadResults();
+
+            log.debug("Send multipart request to S3");
+            amazonS3.completeMultipartUpload(
+                    new CompleteMultipartUploadRequest(
+                            multiPartUploadResult.getBucketName(),
+                            multiPartUploadResult.getKey(),
+                            multiPartUploadResult.getUploadId(),
+                            partETags
+                    )
+            );
+        }
+        catch (ExecutionException e) {
+            abortMultiPartUpload();
+
+            log.error(e.getMessage(), e);
+            throw new IOException("Multi part upload failed ", e.getCause());
+        }
+        catch (InterruptedException e) {
+            log.error(e.getMessage(), e);
+            abortMultiPartUpload();
+            Thread.currentThread().interrupt();
+        }
+        finally {
+            currentOutputStream = null;
+        }
+    }
+
+    private void initiateMultiPartIfNeeded() {
+        if (multiPartUploadResult == null) {
+
+            log.debug("Initiate multi part");
+            multiPartUploadResult = amazonS3
+                    .initiateMultipartUpload(new InitiateMultipartUploadRequest(bucketName, objectName));
+        }
+    }
+
+    private void abortMultiPartUpload() {
+        log.debug("Abort multi part");
+
+        if (isMultiPartUpload()) {
+            amazonS3.abortMultipartUpload(new AbortMultipartUploadRequest(
+                            multiPartUploadResult.getBucketName(),
+                            multiPartUploadResult.getKey(),
+                            multiPartUploadResult.getUploadId()));
+        }
+    }
+
+    private List<PartETag> getMultiPartsUploadResults() throws ExecutionException, InterruptedException {
```

Comment:
We dont use Exceptions in the interface.

Reply:
this is private method which called in finishMultiPartUpload method. Please take a look finishMultiPartUpload method. Here we need exception. It's allow us handle interruption of uploading correctly